## Linear Regression
Linear Regression is a Machine Learning model to predict a line that best fit the given data, 
such that, if new data points are given, we can make prediction based on the line drawn to define dataset.

Basically, we are suppose to find the slope(m) and intercept(b) in a equation of a line: y = mx + b.

NOTE: We have easier method for figure the value of m and b, however, we will perform Gradient Descent.

In [129]:
#Numpy is a python package for scientific computing.
from numpy import *

#### Error Function
Error Function tell us how correct is the value for slope(m) and intercept(b).

Mathematically, Error Function = (y - (m * x + b)). Calculating 'y' as '(m * x + b)' and subtracting it with actual value of 'y'.

The function take the value of slope(m) and intercept(b) to apply to error funcation through all data points and calculating the Average Squared Error for more precise error, and finally, the average error is returned.

In [130]:
def error_function(b, m, dataPoints):
    error = 0
    
    for i in range(0,len(dataPoints)):
        x = dataPoints[i,0]
        y = dataPoints[i,1]
        
        error += (y - (m * x + b))**2
        
    averageError = error / float(len(dataPoints))
    return averageError

#### Gradient Descent
Gradient Descent is a process of finding the best fitting value (local maxima) of slope(m) and intercept(m).

Mathematically, gradient descent is a partial derivative of error function. Therefore,
Partial Derivative w.r.t. slope(m): -(2/n) * (y - (m  * x) + b)
Partial Derivative w.r.t. intercept(m): -(2/n) * x * ((y - (m * x)) + b)

The following function loop through all the datapoints and calculate the gradient for slope and intercept which is later multipled with learning rate and the actual slope and intercept value is adjusted.

In [131]:
def gradient_descent(b, m, dataPoints, learning_rate):
    gradientB = 0
    gradientM = 0
    
    n = float(len(dataPoints))
    
    for i in range (0, len(dataPoints)):
        x = dataPoints[i,0]
        y = dataPoints[i,1]
        
        gradientB += -(2/n) * (y - (m  * x) + b)
        gradientM += -(2/n) * x * ((y - (m * x)) + b)
    
    adjustedB = b - (learning_rate * gradientB)
    adjustedM = m - (learning_rate * gradientM)
    return [adjustedB, adjustedM]

### Training Model

#### Dataset
Given dataset looking like following:

    32.502345269453031,31.70700584656992
    53.426804033275019,68.77759598163891

In above dataset is in 'x,y' format. 

Dataset File: linear-regression-data.csv

In [132]:
#get all data points in (x,y) format
dataPoints = genfromtxt("linear-regression-data.csv", delimiter=",")

#### Hyper-Parameters
- Learning Rate is the stepping size that defines how fast of slow a model will converge.
- Initializing initial value of slope and intercept.
- The number of iteration we want our model to run to given best fitting values for slope and intercept.


In [133]:
learning_rate = 0.0001
iniB = 0
iniM = 0
noIteration = 1000

In [134]:
#Calculate initial error
print("INITIAL ERROR: %f" % error_function(iniB, iniM, dataPoints))

#Iterate modal to train and adjust the value of slope (m) and intercept (b) over each iteration with gradient descent
b = iniB
m = iniM
for i in range(noIteration):
    b, m = gradient_descent(b, m, array(dataPoints), learning_rate)
    
    #Uncomment following line to see training progress
    #print("SLOPE(m): %f | INTERCEPT(b): %f | ERROR: %f" % (b, m, error_function(b, m, dataPoints)))

#Training Result
print("INTERCEPT (b): %f and SLOPE (m): % f" % (b, m))
print("FINAL ERROR: %f" % error_function(b, m, dataPoints))

INITIAL ERROR: 5565.107834
INTERCEPT (b): 0.089899 and SLOPE (m):  1.481254
FINAL ERROR: 112.645300


#### Observations
The initial error (5565.107834) was was reduced to (112.645300).

Now, we have the value for slope(m) and intercept(b), i.e., we have the model (line) that defins this dataset such that if we are given an input feature (x) for this data set, we can predict the output (y).

That's Linear Regression Model in Machine Learning!